Convert directives in components folder to component directives

ankita240796 opened the pull request:
## Explanation
Migrate directives in components folder to component directives.

ankita240796 committed:
Attribution guide

ankita240796 committed:
Background banner

ankita240796 committed:
Create button

ankita240796 committed:
Loading

ankita240796 committed:
Profile link image

ankita240796 committed:
Profile link text

ankita240796 committed:
Promo bar

ankita240796 committed:
Sharing links

ankita240796 committed:
Side navigation bar

ankita240796 committed:
Social buttons

ankita240796 committed:
Top navigation bar

ankita240796 committed:
Fix loading dots

ankita240796 committed:
Circular Image

ankita240796 committed:
Collection summary tile

ankita240796 committed:
Story summary tile

ankita240796 committed:
Hint and Solutions button

ankita240796 committed:
Summary list header

ankita240796 committed:
Version diff visualization

ankita240796 committed:
Fix publication test

ankita240796 committed:
Answer group editor

ankita240796 committed:
Hint editor

ankita240796 committed:
Outcome destination editor

ankita240796 committed:
Outcome editor

ankita240796 committed:
Outcome feedback editor

ankita240796 committed:
Response header

ankita240796 committed:
Rule editor

ankita240796 committed:
Rule type selector

ankita240796 committed:
Solution editor

ankita240796 committed:
Solution explanation editor

ankita240796 committed:
Fix answer group editor

seanlip commented on core/templates/dev/head/components/HintAndSolutionButtonsDirective.ts:
```diff
@@ -21,8 +21,10 @@ oppia.directive('hintAndSolutionButtons', [
     return {
       restrict: 'E',
       scope: {},
+      bindToController: true,
```
Shouldn't this be {}? I think it's mandated -- https://angular.io/guide/upgrade#using-component-directives

ankita240796 replied:
Oh yes, my bad. I was following [this guide](https://toddmotto.com/no-scope-soup-bind-to-controller-angularjs/) as well and used both true & {} as needed. I will make the updates.

ankita240796 replied:
Done.

seanlip replied:
Ah, even in the guide you linked, it looks like making bindToController into an object and moving scope properties across to it is the preferred method, so yeah I think the refactor makes sense.

seanlip commented on core/templates/dev/head/components/HintAndSolutionButtonsDirective.ts:
```diff
@@ -21,8 +21,10 @@ oppia.directive('hintAndSolutionButtons', [
+      bindToController: true,
       templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
         '/components/hint_and_solution_buttons_directive.html'),
+      controllerAs: 'hintAndSolutionButtonsCtrl',
```
Just to check, any reason we don't go for $ctrl like in https://angular.io/guide/upgrade#using-component-directives ?

ankita240796 replied:
No specific reason, I just thought it would be better to name each controller according to the directive. I am fine with both of the ways. Although now I feel $ctrl will be better to have consistency everywhere in all templates. Shall I make the changes for the same?

seanlip replied:
That's your call :)

seanlip replied:
(But please say which one you decide on, and why. If you are not sure, I would advise following standards where possible.)

ankita240796 replied:
I have decided to go with $ctrl since it is more consistent to use the same controllerAs syntax everywhere as mentioned in standards rather than having a new name for each of the controller.

seanlip commented on core/templates/dev/head/components/HintAndSolutionButtonsDirective.ts:
```diff
@@ -21,8 +21,10 @@ oppia.directive('hintAndSolutionButtons', [
         '/components/hint_and_solution_buttons_directive.html'),
+      controllerAs: 'hintAndSolutionButtonsCtrl',
       controller: [
         '$scope', '$rootScope', 'HintsAndSolutionManagerService',
```
Here and elsewhere, should we remove $scope from the dependency list?

ankita240796 replied:
There are some directives where we have `$scope.$on` / `$scope.$apply`. I am not able to find a substitute for this for controller binding. So, this part was left the same. It works correctly but we cannot use $scope in Angular2. So, if we don't get a substitute in AngularJS, the other option will be to make these updates when we are just going to make the final upgrade step. What do you suggest here?
\cc @YashJipkate

seanlip replied:
Ah ok, thanks for the explanation! I am fine with leaving it (I was just curious).

seanlip reviewed (changes requested):
Hi @ankita240796, I just looked at one of these; similar comments apply elsewhere. PTAL?

Thanks!

ankita240796 committed:
Update bindToController

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into comp-dir

ankita240796 committed:
Fix form binding

ankita240796 committed:
Change naming to

seanlip commented:
@ankita240796 -- just a note, let me know when you're ready for another review pass and I'll do one. Thanks!

ankita240796 commented:
Hi @seanlip, the suggested changes are done. You can take another pass now, Thanks!

seanlip commented on core/templates/dev/head/components/side_navigation_bar/SideNavigationBarDirective.ts:
```diff
@@ -21,13 +21,16 @@ oppia.directive('sideNavigationBar', [
-        $scope.getStaticImageUrl = UrlInterpolationService.getStaticImageUrl;
+      controllerAs: '$ctrl',
+      controller: ['$timeout', function(
+          $timeout) {
```
Nit: can probably move to previous line.

ankita240796 replied:
Done.

seanlip commented on core/templates/dev/head/components/loading/LoadingDotsDirective.ts:
```diff
@@ -20,7 +20,11 @@ oppia.directive('loadingDots', [
       templateUrl: UrlInterpolationService.getDirectiveTemplateUrl(
-        '/components/loading/loading_dots_directive.html')
+        '/components/loading/loading_dots_directive.html'),
+      controllerAs: '$ctrl',
```
Just to check -- in cases like this, do we even need the controllerAs and controller fields?

ankita240796 replied:
Yes, it is mandatory for any component directive to have all these fields as specified in the guide [here](https://angular.io/guide/upgrade#using-component-directives).

seanlip commented on core/templates/dev/head/components/side_navigation_bar/SideNavigationBarDirective.ts:
```diff
@@ -21,13 +21,16 @@ oppia.directive('sideNavigationBar', [
+      controller: ['$timeout', function(
+          $timeout) {
+        var ctrl = this;
+        ctrl.currentUrl = window.location.pathname;
```
Maybe we should do this separately but just a note that this should probably either make use of $window rather than window, or use UrlService perhaps.

ankita240796 replied:
Filed an issue #6738

seanlip reviewed (approved):
Hi @ankita240796, generally LGTM. Thanks very much!

One question -- how are you going to enforce these formats? I was expecting to see updates to the lint checks to ensure that devs define the right properties on the directives etc., otherwise we'll see regressions. Or is there an Angular eslint rule we can turn on?

ankita240796 committed:
Remove extra line

ankita240796 commented:
Hi @seanlip, I have addressed the comments.

> Hi @ankita240796, generally LGTM. Thanks very much!

You're welcome :)

> One question -- how are you going to enforce these formats? I was expecting to see updates to the lint checks to ensure that devs define the right properties on the directives etc., otherwise we'll see regressions. Or is there an Angular eslint rule we can turn on?

We cannot enforce this until all directives are changed to component directives. This will be only possible before the final upgrade because there is no substitute for $compile in AngularJS. We will use dynamic loading only when we are making the final upgrade to Angular2 and then $compile, $scope left out now can be removed. 

As of now, I think we will have to update the wiki and be more careful during reviews to ensure that violations do not happen.

seanlip commented:
OK, LGTM (and approval) from my end. Thanks!

@YashJipkate is the check I mentioned necessary (and is it part of your project)?

YashJipkate commented:
Hi @seanlip Yes the lint checks would be useful and helpful to the devs. But the checks can only be employed once all the controllers and directives are refactored. Once they are done, we will have a lint check as with the modularity stuff.

Thanks

ankita240796 committed:
Fix regex

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into comp-dir

seanlip commented:
@YashJipkate, can you review this? I'm comfortable merging once you have, since this is a codebase refactoring change. Thanks!

YashJipkate commented on core/templates/dev/head/components/summary_tile/CircularImageDirective.ts:
```diff
@@ -20,15 +20,18 @@ oppia.directive('circularImage', [
+      controller: [function() {
+        var ctrl = this;
+        this.isLinkAvailable = function() {
+          return this.link() ? true : false;
```
Why `this` instead of `ctrl`?

ankita240796 replied:
My bad, Fixed.

YashJipkate commented on core/tests/protractor_utils/general.js:
```diff
@@ -143,7 +143,9 @@ var ensurePageHasNoTranslationIds = function() {
-        new RegExp('ng-show="navElementsVisibilityStatus.I18N_', 'g');
+        new RegExp(
+          'ng-show="\\$ctrl.navElementsVisibilityStatus.I18N_',
+          'g');
```
Do we really need to break this up?

ankita240796 replied:
Fixed.

YashJipkate reviewed (changes requested):
LGTM Just some queries. PTAL!

ankita240796 committed:
Remove extra lines

ankita240796 committed:
Replace this with ctrl

ankita240796 commented:
Hi @YashJipkate, I have addressed the review comments, PTAL. Thanks!

YashJipkate reviewed (approved):
LGTM! @ankita240796 
Thanks!

seanlip commented:
Merged!
